Implement character/key conversion

niksilver opened the pull request:
As discussed in #1657, enable developers to work with printable characters from within `keyboard.code()`, plus supporting documentation.
- Includes new function `keyboard.code_to_char()`. 
- Includes a new function `keyboard.char_to_code()`.
- Replaced some code from `keyboard.process()`, to use `code_to_char()` - which just uses that original code.

I suggest further discussion on this topic takes place here, not in the originating issue.

This PR does not include updates to the `docs/` directory because when I generated the docs it made too many changes (most notably updating every external reference in the docs to the Lua 5.3 manual). I presume the docs will be generated before an official release. However, I have tested locally using `ldoc`.

This PR does not include the suggested `keyboard.char_lift()` as it duplicates functionality (a bit) and I thought it might just be too much.

I have tested this manually using [the script over here](https://github.com/niksilver/norns-kcode-test) which can be installed via the maiden REPL using
```
;install https://github.com/niksilver/norns-kcode-test
```

p3r7 commented:
`code_to_char` implem is looking mighty fine.

on the other hand i'm not convinced by the usefulness of `char_to_code` nor its robustness.

don't get me wrong. it's not a issue of code quality but the problem is that `code_to_char` is **certainly not bijective** (not a 1 to 1 code / char mapping)!
(also `ikm` should be recomputed on change of `keyboard.selected_map`, inside of `keyboard.load_enabled()`)

niksilver commented:
On your points, well spotted on these two:
* Yes, `char_to_code` is certainly not bijective. I meant to add something to that effect in the documentation, but forgot.
* Yes, `ikm` should be recomputed. That error is due to laziness on my part. 😞 

As to the usefulness of `char_to_code`... I don't know. Mainly it seemed it might a reasonable countpart to `code_to_char`. It might be used in the situation where the developer wants to test for a character in `keyboard.code()` because that's where the rest of their logic is. Eg:

```lua
QUIT_BUTTON = keyboard.char_to_code('Q')

function keyboard.code(code, value)
  if code == QUIT_BUTTON and value > 0 then
    do_quit()
  end
  -- etc
end
```

That seems a bit nicer than this:
```lua
function keyboard.code(code, value)
  local is_q = keyboard.code_to_char(code) == 'Q' or keyboard.code_to_char(code) == 'q')
  if is_q and value > 0 then
    do_quit()
  end
  -- etc
end
```

Meanwhile this is simple, but only if the developer is happy to use `keyboard.char()` - which might not be the case if the rest of their logic is in `keyboard.code()`:
```lua
function keyboard.char(ch)
  if ch == 'Q' or ch == 'q' then
    do_quit()
  end
  -- etc
end
```

So... I think it has a bit of utility. But I don't feel strongly about it.

p3r7 commented:
the issue w/ `char_to_code` not being bijective is not a documentation one.

it's that `code_to_char` is not guaranteed to work if several key combination in a layout can result in the same char. in that case a simple map for reverse lookup may not cut it.

likewise, `char_to_code` would have to take into account current modifiers (like `char_to_code`). untested but i predict that w/ current implem if i try to map the `QUIT_BUTTON` to say the `0` on my AZERTY keyboard, it'd get shadowed by either `@` or `à` as they all share code but w/ different modifiers.

the `keyboard.code_to_char` approach may be more verbose but is more correct.

niksilver commented:
I've now removed code relating to `char_to_code()` and left only `code_to_char()` with the updated documentation. I've also updated my [manual testing script](https://github.com/niksilver/norns-kcode-test).

So this is ready for further review.

tehn commented:
apologies for the delay. any further thoughts on this one? @niksilver @p3r7

p3r7 commented:
everything looks good to me.

niksilver commented:
Yes, I'm happy with what we have here.
